test/cbmc/sources/mqtt_agent_cbmc_state.c

```diff
@@ -0,0 +1,139 @@
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file mqtt_agent_cbmc_state.c
+ * @brief Implements the functions defined in mqtt_agent_cbmc_state.h.
+ */
+#include <stdint.h>
+#include <stdlib.h>
+#include "mqtt_agent.h"
+#include "mqtt_agent_cbmc_state.h"
+#include "network_interface_stubs.h"
+#include "get_time_stub.h"
+#include "incoming_publish_callback_stub.h"
+#include "agent_message_stubs.h"
+#include "agent_command_pool_stubs.h"
+
+
+MQTTFixedBuffer_t * allocateMqttFixedBuffer( MQTTFixedBuffer_t * pFixedBuffer )
+{
+    if( pFixedBuffer == NULL )
+    {
+        pFixedBuffer = malloc( sizeof( MQTTFixedBuffer_t ) );
+    }
+
+    if( pFixedBuffer != NULL )
+    {
+        __CPROVER_assume( pFixedBuffer->size < CBMC_MAX_OBJECT_SIZE );
+        pFixedBuffer->pBuffer = malloc( pFixedBuffer->size );
+    }
+
+    return pFixedBuffer;
+}
+
+bool isValidMqttFixedBuffer( const MQTTFixedBuffer_t * pFixedBuffer )
+{
+    bool isValid = true;
+
+    if( pFixedBuffer != NULL )
+    {
+        isValid = pFixedBuffer->size < CBMC_MAX_OBJECT_SIZE;
+    }
+
+    return isValid;
+}
+
+MQTTAgentContext_t * allocateMqttAgentContext( MQTTAgentContext_t * pContext )
+{
+    TransportInterface_t * pTransportInterface;
+    MQTTFixedBuffer_t * pNetworkBuffer;
+    AgentMessageInterface_t * pMessageInterface;
+    MQTTStatus_t status = MQTTSuccess;
+    void * pIncomingCallbackContext;
+
+    if( pContext == NULL )
+    {
+        pContext = malloc( sizeof( MQTTAgentContext_t ) );
+    }
+
+    pTransportInterface = malloc( sizeof( TransportInterface_t ) );
+
+    if( pTransportInterface != NULL )
+    {
+        /* The possibility that recv and send callbacks are NULL is tested in the
+         * MQTT_Init proof. MQTT_Init is required to be called before any other
+         * function in core_mqtt.h. */
+        pTransportInterface->recv = NetworkInterfaceReceiveStub;
+        pTransportInterface->send = NetworkInterfaceSendStub;
+    }
+
+    pNetworkBuffer = allocateMqttFixedBuffer( NULL );
+    __CPROVER_assume( isValidMqttFixedBuffer( pNetworkBuffer ) );
+
+    pMessageInterface = malloc( sizeof( AgentMessageInterface_t ) );
+
+    if( pMessageInterface != NULL )
+    {
+        pMessageInterface->send = AgentMessageSendStub;
+        pMessageInterface->recv = AgentMessageRecvStub;
```

Comment:
I think these stubs are missing their definition

Reply:
Yea that was an intentional choice. I just kept the declaration of the function so that the return will always be an no deterministic value from either true or false( return type of the function). It will be just enough for the proofs.

Reply:
It might still be better to define a stub that returns `nondet_bool()` so that it is self-documenting.

Reply:
Agree. Makes sense. I will add that